trio/tests/test_util.py

```diff
@@ -58,3 +61,54 @@ def handler(signum, _):
                 async with ul1:
                     pass  # pragma: no cover
     assert "ul1" in str(excinfo.value)
+
+
+async def test_contextmanager_do_not_unchain_non_stopiteration_exceptions():
+    @acontextmanager
+    @async_generator
+    async def test_issue29692():
+        try:
+            await yield_()
+        except Exception as exc:
+            raise RuntimeError('issue29692:Chained') from exc
```

Comment:
OKAY, I figured out the problem with `StopIteration` and `async_generator`!

Here it is: when a `StopIteration` gets thrown into an `async_generator`, it actually ends up inside the magic `await yield_()` function. But Python doesn't know that it's supposed to be a magic function that acts like `yield`; it just thinks its some random async function. And then it tries to raise `StopIteration`, and Python's like uh, no, haven't you heard of PEP 479? Async functions can't raise `StopIteration`. So it immediately wraps it into a `RuntimeError`. Which ... is a little weird and annoying, but there's nothing we can do about it, and it doesn't stop us from unwrapped it later. Except... the way this particular test is written, where it wraps everything in a `RuntimeError`, means that we end up wrapping our `StopIteration` in a `RuntimeError` *twice*. We get `RuntimeError` → `RuntimeError` → `StopIteration`.

So there is something of a bug here, but it's intrinsic to how `async_generator` works, and we can't fix that part.

But! We can still test the main thing we care about, which is that a `StopIteration` is able to pass through a context manager like
```python
@acontextmanager
@async_generator
async def simpler_context_manager():
    await yield_()
```

And this should work, though slightly differently than if it were a native generator. With a native generator, the `yield` would raise `StopIteration`, which would then get wrapped in a `RuntimeError` when it tried to leave the function. With `async_generator`, the `await yield_()` directly wraps the `StopIteration` into a `RuntimeError`, and then that propagates through the function and out. But either way we can catch it and unwrap it in `@acontextmanager`.